### ISSUE TITLE
[Popup] Define accessibility mappings for popup and its primitives

### ISSUE THREAD

**Issue by melanierichards** [openui/open-ui#329]
Typically, when new primitives/elements are added to the platform, stakeholders of web plat accessibility will work with the editors of the HTML AAM to define accessibility mappings. This step should come a little later in the process when the design of said primitives is more formalized. Tracking this need here, as there has been express author interest in this.

In scope for this new mapping definition, currently:

* `popup` element
* `popup` attribute
* `anchor` attribute
* `delegatesfocus` attribute
* Possibly something like an `initiallyopen` attribute (refer to https://github.com/WICG/open-ui/issues/311)

**Comment by scottaohara** [comment]
a) yes, the element that has the `popup` attribute will be the element to convey the necessary role, and since popup is no longer an element, if someone needed to create a custom component (e.g., with a `div` or `custom-element`) then an ARIA role could be assigned as necessary.

if there are any other accessibility semantics that the `popup` attribute itself would communicate, https://github.com/w3c/aria/issues/1697 is the issue I opened with ARIA to define those.  I have thoughts, if you need something sooner than later.

Regarding the trigger element, this is where the type of popup will be important.  Because a default popup _may_ need `aria-haspopup` or `aria-expanded` (there is discussion going on in ARIA concerning the actual usefulness of the `aria-haspopup` attribute and a bunch of baggage associated with it that I won't bore you with right now.  

If a `popup=hint` then it would be assumed that an `aria-describedby` association would be created with the tooltip popup - no need for an aria-expanded or aria-haspopup. 

If it's an async popup, e.g., a toast notification popup - then we likely wouldn't want any mapping for the attribute as the live region announcement (or ideally exposed via a notification api - https://github.com/w3c/aria/issues/832)  of the popup's contents would be the audio cue of the popup... popping up.

Please let me know if you have more questions / feel free to reach out and we can discuss these further if need be, especially if you are blocked on anything.

**Comment by mfreed7** [comment]
> a) yes, the element that has the `popup` attribute will be the element to convey the necessary role, and since popup is no longer an element, if someone needed to create a custom component (e.g., with a `div` or `custom-element`) then an ARIA role could be assigned as necessary.
> 
> if there are any other accessibility semantics that the `popup` attribute itself would communicate, [w3c/aria#1697](https://github.com/w3c/aria/issues/1697) is the issue I opened with ARIA to define those. I have thoughts, if you need something sooner than later.

Ok great, thanks for confirming. If there are conclusions that come out of https://github.com/w3c/aria/issues/1697 for the popup itself, definitely <span>@</span>mention me so I can implement them.


> Regarding the trigger element, this is where the type of popup will be important. Because a default popup _may_ need `aria-haspopup` or `aria-expanded` (there is discussion going on in ARIA concerning the actual usefulness of the `aria-haspopup` attribute and a bunch of baggage associated with it that I won't bore you with right now.
> 
> If a `popup=hint` then it would be assumed that an `aria-describedby` association would be created with the tooltip popup - no need for an aria-expanded or aria-haspopup.
> 
> If it's an async popup, e.g., a toast notification popup - then we likely wouldn't want any mapping for the attribute as the live region announcement (or ideally exposed via a notification api - [w3c/aria#832](https://github.com/w3c/aria/issues/832)) of the popup's contents would be the audio cue of the popup... popping up.

Ok, to be sure I understood this correctly, for the trigger element:
- Let's call the element (likely a popup) pointed to by the triggering attribute (e.g. `togglepopup`) the *"target"*.
- If the *target* isn't a valid popup, do nothing.
- If the *target* has `popup=popup`, then set `aria-haspopup=true` on the trigger element. Also set `aria-expanded` to true if/when the *target* popup is showing, and false when it is hidden.
- If the *target* has `popup=hint`, then set `aria-describedby` to the *target*.
- If the *target* has `popup=async`, then do nothing.
- In all of the above cases, when I say "set `aria-*`", I mean in the accessibility tree only. I do not mean the element should sprout content attributes.

Any mistakes there?

**Comment by scottaohara** [comment]
re bullet 3 i would actually say at this point `aria-expanded=false` when the popup is not shown, and `aria-expanded=true` when it is.  I really don't know at this point whether ARIA will be using `aria-haspopup` going forward or deprecating that attribute.  i'll re-sync with a few people on that, but in the meantime the conversation was going towards using `aria-expanded` instead.  but there are some bugs with `aria-haspopup` when used on a button... it always makes the button expose as a 'menu button' with a 'menu popup' - regardless of the haspopup value... so hence my caginess on you actually implementing that feature at this time.  

it's safe for something like a combobox with listbox popup, but just rubbish for buttons at this point.  :(

**Comment by scottaohara** [comment]
another thought that came to mind concerning the hint popup is how presently a `title` attribute can provide certain elements its accessible name if other preferred methods of naming are not provided.

e.g., `<button title=foo></button>` would produce the title tooltip, and the name of the button would be 'foo'.

A hint popup should probably also contribute to the name of the element if the element is otherwise not named.

**Comment by scottaohara** [comment]
some additional accessibility related todos:

* when a generic element is specified to be a `popup`, we would remap that element to be exposed as a `role=group`, rather than a `role=generic`.  This would be defined in HTML AAM.
* Ideally, an additional role description of "popup" would be appended to the implicit, or explicit ARIA role of the element that has been made into a popup.  e.g., `<div popup=popup>`, per the above point, would ideally be exposed to AT in a way so that it could be surfaced as a "group popup".  Similarly a `<dialog popup>`, `<article popup>` or `<div role=note popup>` would potentially be surfaced as "dialog popup", "article popup" and "note popup", respectively.
* for use cases where authors may use a hint popup as a way to expose the label for a button (e.g., a button visually represented by an icon alone), we would also want to ensure that the hint popup could participate in the accName calculation for such elements.  Using hyperlink naming as an example: For an `<a href>` element:
  - Use `aria-labelledby` or `aria-label`
  - Otherwise, use subtree of the hyperlink
  - Otherwise, use the subtree (accName?) of the `popup=hint` element associated with the hyperlink.
  - Otherwise, use the `title` attribute's value
  - If none of the above are specified, the hyperlink has no accessible name.

cc @aleventhal @mfreed7

**Comment by aleventhal** [comment]
"No accessible content" — would need a definition.
We might want to treat it as aria-hidden="true" as opposed to role="none".

**Comment by aleventhal** [comment]
More thoughts on a11y semantics:
* Consider putting in landmark cycle, along with non-modal dialogs (there's an ongoing discussion about this)
* Consider using role=alertdialog rather than role=dialog.
* Consider automatically adding aria-live="polite" and letting author override. Potentially helpful, but also potentially too verbose at times.
* Consider a mechanism to expose that it is a popup, since the role could be anything (perhaps object attribute ispopup:true for platforms that support object attributes).
* Consider firing dialog start event (exists only in MSAA)

F6 for navigation will be helpful. Another thing to consider is whether the Escape key can close it, as it could in a non-modal dialog. It doesn't seem like that's a possibility here, since we don't know if the author wants the user to be allowed to close it.

**Comment by scottaohara** [comment]
>Another thing to consider is whether the Escape key can close it, as it could in a non-modal dialog. It doesn't seem like that's a possibility here, since we don't know if the author wants the user to be allowed to close it.

this seems like a reasonable feature so long as the popup can receive focus / contains an element that can receive focus.  i'm not sure if this is something that authors would always want, but so long as there is a way to override that would seem fine to me?  also, maybe makes sense to have this rule apply to `popup=hint` as well, since, if that can have a focusable element it too should be able to be dismissed via <kbd>esc</kbd>.

Regarding making `popup=async` have live region semantics by default... whether that be `alertdialog` or `aria-live=polite` - agreed... i'm concerned with the amount of chatter this would cause, as well as the potential of individual popups that fire at the same time overwriting each other, or them just not announcing at all per present AT issues with live regions that get injected into web pages.  there could be multiple `popup=async` on page load, which could also be overwhelming if they were all announced at once.  definitely lots to think about here still

**Comment by aleventhal** [comment]
Mason, what's your opinion about the escape key inside of a popup=async
that actually takes focus?

Scott, let's track the live region question in a separate GitHub issue
since it's going to require real community discussion.
(I'm concerned that if we put too many things in the same GitHub issues the
conversation will get confusing.
So, especially for potentially controversial tradeoffs, we should put them
in their own issues ... right?)

Aaron

**Comment by mfreed7** [comment]
Ok, per the [resolution on 495](https://github.com/openui/open-ui/issues/495#issuecomment-1164827851) today, all of this great a11y discussion needs to be moved to the ARIA working group, to determine how to handle the popup attribute correctly.

As such, I'd like to close this issue. So as to avoid losing the context from this discussion, please link back to this issue from the ARIA issue, once it's been opened. Thanks!
